cggmp21/src/signing.rs

```diff
@@ -1160,68 +1273,68 @@ where
-        let enc_j = fast_paillier::EncryptionKey::from_n(R_j.N.clone());
-
-        let data = pi_log::Data {
-            key0: &enc_j,
-            c: &ciphertext_j.K,
-            x: &msg_j.Delta,
-            b: &Gamma,
-        };
-
-        if pi_log::non_interactive::verify::<E, D>(
-            &unambiguous::ProofLog {
-                sid,
-                prover: j,
-                prime_prime: true,
-            },
-            &R_i.into(),
-            data,
-            &msg_j.psi_prime_prime.0,
-            &security_params.pi_log,
-            &msg_j.psi_prime_prime.1,
-        )
-        .is_err()
-        {
-            faulty_parties.push((j, ciphertext_id, msg_id))
-        }
-    }
-    runtime.yield_now().await;
-
+    let faulty_parties =
+        utils::collect_blame_with_data(&round3_msgs, &ciphertexts, |j, msg_j, ciphertext_j| {
+            pi_elog::non_interactive::verify::<E, D>(
+                &unambiguous::ProofElog {
+                    sid,
+                    prover: j,
+                    prime: true,
+                },
+                pi_elog::Data {
+                    l: &ciphertext_j.A1,
+                    m: &ciphertext_j.A2,
+                    x: &ciphertext_j.Y,
+                    y: &msg_j.Delta,
+                    h: &Gamma,
+                },
+                &msg_j.psi_prime.0,
+                &msg_j.psi_prime.1,
+            )
+            .err()
+        });
     if !faulty_parties.is_empty() {
         return Err(SigningAborted::InvalidPsiPrimePrime(faulty_parties).into());
     }
+    runtime.yield_now().await;
 
     // Step 2
     tracer.stage("Calculate presignature");
     let delta = delta_i + round3_msgs.iter().map(|m| m.delta).sum::<Scalar<E>>();
     let Delta = Delta_i + round3_msgs.iter().map(|m| m.Delta).sum::<Point<E>>();
+    let S = S_i + round3_msgs.iter().map(|m| m.S).sum::<Point<E>>();
 
-    if Point::generator() * delta != Delta {
+    if Point::generator() * delta != Delta || pk * delta != S {
         // Following the protocol, party should broadcast additional proofs
         // to convince others it didn't cheat. However, since identifiable
         // abort is not implemented yet, this part of the protocol is missing
         return Err(SigningAborted::MismatchedDelta.into());
     }
 
-    let R = Gamma * delta.invert().ok_or(Bug::ZeroDelta)?;
-    let R = NonZero::from_point(R).ok_or(Bug::ZeroR)?;
+    let delta_inv = delta.invert().ok_or(Bug::Zero(BugSource::delta))?;
+
+    let tilde_k_i = SecretScalar::new(&mut (k_i * delta_inv));
+    let tilde_chi_i = SecretScalar::new(&mut (chi_i * delta_inv));
+
+    let Gamma = NonZero::from_point(Gamma).ok_or(Bug::Zero(BugSource::Gamma))?;
+
+    let commitments = round3_msgs
+        .iter_including_me(&my_round3_msg)
+        .map(|m| (delta_inv * m.Delta, delta_inv * m.S))
+        .collect::<Vec<_>>();
+    let commitments = PresignatureCommitments { Gamma, commitments };
```

Comment:
Should it be a proper struct instead of a tuple for commitments?

Reply:
Yes, that's what I did

Reply:
What about tuples though? Do you suggest any name for a structure?